akka-stream-tests/src/test/scala/akka/stream/scaladsl/SetupSpec.scala

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import akka.NotUsed
+import akka.stream.ActorMaterializer
+import akka.stream.testkit.StreamSpec
+
+class SetupSpec extends StreamSpec {
+
+  implicit val materializer = ActorMaterializer()
+  import system.dispatcher
+
+  "Source.setup" should {
+
+    "expose materializer" in {
+      val source = Source.setup { mat ⇒ _ ⇒
+        Source.single(mat.isShutdown)
+      }
+
+      source.runWith(Sink.head).futureValue shouldBe false
+    }
+
+    "expose attributes" in {
+      val source = Source.setup { _ ⇒ attr ⇒
+        Source.single(attr.attributeList)
+      }
+
+      source.runWith(Sink.head).futureValue should not be empty
+    }
+
+    "propagate materialized value" in {
+      val source = Source.setup { _ ⇒ _ ⇒
+        Source.maybe[NotUsed]
+      }
+
+      val (completion, element) = source.toMat(Sink.head)(Keep.both).run()
+      completion.futureValue.trySuccess(Some(NotUsed))
+      element.futureValue shouldBe NotUsed
+    }
+
+    "propagate attributes" in {
+      val source = Source.setup { _ ⇒ attr ⇒
+        Source.single(attr.nameLifted)
+      }.named("my-name")
+
+      source.runWith(Sink.head).futureValue shouldBe Some("my-name")
+    }
+
+    "propagate attributes when nested" in {
+      val source = Source.setup { _ ⇒ _ ⇒
+        Source.setup { _ ⇒ attr ⇒
+          Source.single(attr.nameLifted)
+        }
+      }.named("my-name")
+
+      source.runWith(Sink.head).futureValue shouldBe Some("my-name")
+    }
+
+    "handle factory failure" in {
+      val error = new Error("boom")
+      val source = Source.setup { _ ⇒ _ ⇒
+        throw error
+      }
+
+      val (materialized, completion) = source.toMat(Sink.head)(Keep.both).run()
+      materialized.failed.futureValue.getCause shouldBe error
+      completion.failed.futureValue.getCause shouldBe error
+    }
+
+    "handle materialization failure" in {
+      val error = new Error("boom")
+      val source = Source.setup { _ ⇒ _ ⇒
+        Source.empty.mapMaterializedValue(_ ⇒ throw error)
+      }
+
+      val (materialized, completion) = source.toMat(Sink.head)(Keep.both).run()
+      materialized.failed.futureValue.getCause shouldBe error
+      completion.failed.futureValue.getCause shouldBe error
+    }
+
+  }
+
+  "Flow.setup" should {
+
+    "expose materializer" in {
+      val flow = Flow.setup { mat ⇒ _ ⇒
```

Comment:
I think it would be easier to understand the API if it was `(mat, att) =>` instead of function of function

Reply:
@2m @ennru Did you see the feedback here?

Reply:
Thanks for the ping. Yea, `(mat, att) =>` is more conventional, but then we loose ability to specify that the parameter is implicit, which is very useful for materializer, like `implicit mat => attr =>`.

I agree, that trading that for more conventional API and similarity to Java API is better. I'll update the PR.